pkg/destroy/azure/azure.go

```diff
@@ -184,37 +194,74 @@ func (o *ClusterUninstaller) Run() error {
 }
 
 func deletePublicRecords(ctx context.Context, dnsClient dns.ZonesClient, recordsClient dns.RecordSetsClient, privateDNSClient privatedns.PrivateZonesClient, privateRecordsClient privatedns.RecordSetsClient, logger logrus.FieldLogger, rgName string) error {
-	ctx, cancel := context.WithTimeout(ctx, 10*time.Minute)
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Minute)
 	defer cancel()
 
 	// collect records from private zones in rgName
 	var errs []error
-	for zonesPage, err := dnsClient.ListByResourceGroup(ctx, rgName, to.Int32Ptr(100)); zonesPage.NotDone(); err = zonesPage.NextWithContext(ctx) {
+
+	for {
+		zonesPage, err := dnsClient.ListByResourceGroup(ctx, rgName, to.Int32Ptr(100))
```

Comment:
why this change? Previous code also checked and caught all the errors?

Reply:
It turns out that the previous code hid the errors, this is why I changed it to be this way.

Reply:
Can you expand on when it would hide errors?

Reply:
Any error returned by dnsClient.ListByResourceGroup() is hidden. I'm not exactly clear on why, but I attempted a number of ways to rewrite this loop, and all of them mask any errors. This way is a compromise.

Reply:
For example, editing osServicePrincipal.json and making the tenant id invalid, should cause an error here, but it gets masked. Easily testable.

Reply:
hmm.. so you are saying that the first error from `zonesPage, err := dnsClient.ListByResourceGroup(ctx, rgName, to.Int32Ptr(100));`  is getting hidden.

I tried to create a small test in https://play.golang.org/p/QuKz6Htgx6X and i don't see any error getting hidden.

Reply:
Try this:

<pre>
[jhixson@redlap installer]$ git diff
diff --git a/pkg/destroy/azure/azure.go b/pkg/destroy/azure/azure.go
index 20c0004..25de100 100644
--- a/pkg/destroy/azure/azure.go
+++ b/pkg/destroy/azure/azure.go
@@ -4,6 +4,7 @@ import (
        "context"
        "fmt"
        "net/http"
+       "os"
        "sort"
        "strings"
        "time"
@@ -124,6 +125,7 @@ func (o *ClusterUninstaller) Run() error {
                },
                1*time.Second,
        )
+       os.Exit(1)
        err = waitCtx.Err()
        if err != nil && err != context.Canceled {
                errs = append(errs, errors.Wrap(err, "failed to delete public DNS records"))
@@ -190,19 +192,23 @@ func deletePublicRecords(ctx context.Context, dnsClient dns.ZonesClient, records
        // collect records from private zones in rgName
        var errs []error
        for zonesPage, err := dnsClient.ListByResourceGroup(ctx, rgName, to.Int32Ptr(100)); zonesPage.NotDone(); err = zonesPage.NextWithContext(ctx) {
+               fmt.Printf("XXX: err=%s\n", err)
                if err != nil {
                        errs = append(errs, errors.Wrap(err, "failed to list dns zone"))
                        continue
                }
-               for _, zone := range zonesPage.Values() {
-                       if zone.ZoneType == dns.Private {
-                               if err := deletePublicRecordsForZone(ctx, dnsClient, recordsClient, logger, rgName, to.String(zone.Name)); err != nil {
-                                       errs = append(errs, errors.Wrapf(err, "failed to delete public records for %s", to.String(zone.Name)))
-                                       continue
+               /*
+                       for _, zone := range zonesPage.Values() {
+                               if zone.ZoneType == dns.Private {
+                                       if err := deletePublicRecordsForZone(ctx, dnsClient, recordsClient, logger, rgName, to.String(zone.Name)); err != nil {
+                                               errs = append(errs, errors.Wrapf(err, "failed to delete public records for %s", to.String(zone.Name)))
+                                               continue
+                                       }
                                }
                        }
-               }
+               */
        }
+       os.Exit(1)
        for zonesPage, err := privateDNSClient.ListByResourceGroup(ctx, rgName, to.Int32Ptr(100)); zonesPage.NotDone(); err = zonesPage.NextWithContext(ctx) {
                if err != nil {
                        errs = append(errs, errors.Wrap(err, "failed to list private dns zone"))

</pre>

Edit your ~/.azure/osServicePrincipal.json file to have an incorrect tenantId. This should cause dnsClient.ListByResourceGroup(...) to fail. The inner part of the loop is never reached, the error is never displayed. I've tried adding a second error variable for zonesPage.NextWithContext() as well as testing err as nil, non nil with zonesPage.NotDone() to no success. As written in this PR, the error works as expected.

Reply:
so, zonesPage.NotDone() is the culprit here.

Reply:
If zonesPage.NotDone() is removed, the desired result is achieved. Looks like if the conditional section of the for loop fails, then it's never entered, which sounds about right ;-)

Reply:
@jhixson74  looking closer this change is wrong because,
Previously we call listzones once and each iteration called zonesPage.NextWithContext to go to the next page. but now each iteration call the listzones and then next and then the next iteration calls those again. If there were more than on page we would looping forever.. :P

Reply:
I think the best thing would be

```
page, err := listzones
if err != nil 
   if err not authorized return here
   append the err to errs

for ; page.NotDone; err = page.Next {
    // same as before
}
```